external-crates/move/crates/move-compiler/src/shared/known_attributes.rs

```diff
@@ -0,0 +1,370 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use once_cell::sync::Lazy;
+use std::{collections::BTreeSet, fmt};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum AttributePosition {
+    AddressBlock,
+    Module,
+    Use,
+    Friend,
+    Constant,
+    Struct,
+    Function,
+    Spec,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum KnownAttribute {
+    Testing(TestingAttribute),
+    Verification(VerificationAttribute),
+    Native(NativeAttribute),
+    Diagnostic(DiagnosticAttribute),
+    DefinesPrimitive(DefinesPrimitive),
+    External(ExternalAttribute),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TestingAttribute {
+    // Can be called by other testing code, and included in compilation in test mode
+    TestOnly,
+    // Is a test that will be run
+    Test,
+    // This test is expected to fail
+    ExpectedFailure,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum VerificationAttribute {
+    // deprecated spec only annotation
+    VerifyOnly,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum NativeAttribute {
+    // It is a fake native function that actually compiles to a bytecode instruction
+    BytecodeInstruction,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum DiagnosticAttribute {
+    Allow,
+    // Deprecated lint allow syntax
+    LintAllow,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct DefinesPrimitive;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ExternalAttribute;
+
+impl AttributePosition {
+    const ALL: &'static [Self] = &[
+        Self::AddressBlock,
+        Self::Module,
+        Self::Use,
+        Self::Friend,
+        Self::Constant,
+        Self::Struct,
+        Self::Function,
+        Self::Spec,
+    ];
+}
+
+impl KnownAttribute {
+    pub fn resolve(attribute_str: impl AsRef<str>) -> Option<Self> {
```

Comment:
There are a few places you call this. and then match on the result, discarding most. Would it make sense to write resolves for the individual enums so that we can say, e,g., 

```
attrs.value.iter().filter_map(|attr| {
        TestingAttribute::resolve(attr.value.attribute_name().value)
            .map(|test_attr| (attr.loc, test_attr))?
    });
```

Reply:
It would yeah, but we would still need this one? Or those individual resolves would jsut call this one and map out of it?